### PR TITLE
add package install test, add labels to script steps

### PR DESCRIPTION
**Pull request by kuisathaverat**
* add package install test

Those are moved to other PRs

* add labels to script steps - elastic/apm-server#2378
* add Tool Docker container with the Golang version used and tools needed - elastic/apm-server#2375
* use plain Go or mage commands inside the Tool Docker container - elastic/apm-server#2375
* fix Release build trigger, it was trigger always not only with beats updates. - elastic/apm-server#2379
* move benchmark stage to run in the tool Docker container - elastic/apm-server#2391

### CODE REVIEW

**Comment by graphaelli** [comment]
lgtm, I'll leave the go not in path issue to you

**Comment by kuisathaverat** [comment]
I am checking the APM-Server builds and the build step on linux it is silence failing

**Comment by kuisathaverat** [comment]
there is an issue expanding the path `-files /var/lib/jenkins/workspace/pm-server_apm-server-mbp_PR-2268/src/github.com/elastic/apm-server/build/distributions/*` when we use `mage` the `*` is not expanded to all possible file names.

**Comment by kuisathaverat** [comment]
I was searching for the error in the wrong place, I was thinking that the test does not run because there was no output, it is running but without output, the condition to show the test output was inverted

**Comment by kuisathaverat** [comment]
The test fails with no x86 packages because we cannot test them in an x86 Docker container

```
[2019-06-19T21:40:25.267Z] The command '/bin/sh -c rpm -ivh $apm_server_pkg' returned a non-zero code: 1
[2019-06-19T21:40:25.267Z] --- FAIL: TestRpm (48.69s)
[2019-06-19T21:40:25.267Z]     package_test.go:62: running "docker build --no-cache -q -f Dockerfile.rpm.install --build-arg apm_server_pkg=apm-server-8.0.0-SNAPSHOT-armhfp.rpm /var/lib/jenkins/workspace/pm-server_apm-server-mbp_PR-2268/src/github.com/elastic/apm-server/build/distributions/" failed with exit code 1
[2019-06-19T21:40:25.267Z] FAIL
[2019-06-19T21:40:25.267Z] FAIL	command-line-arguments	191.956s
[2019-06-19T21:40:25.267Z] Error: running "go test -v tests/packaging/package_test.go -files /var/lib/jenkins/workspace/pm-server_apm-server-mbp_PR-2268/src/github.com/elastic/apm-server/build/distributions/* -tags=package" failed with exit code 1
script returned exit code 1
```

**Comment by kuisathaverat** [comment]
@gil should we test only x86 packages?

**Comment by graphaelli** [comment]
That would be a fine start

**Comment by kuisathaverat** [comment]
ready to merge https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp/detail/PR-2268/51/pipeline/191/

File: tests/packaging/package_test.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ var (
 )
 
 func TestDeb(t *testing.T) {
-	debs := getFiles(t, regexp.MustCompile(`\.deb$`))
+	debs := getFiles(t, regexp.MustCompile(`(i386|i686|amd64|x86)+.*\.deb$`))
```
Comment (graphaelli) [inline tests/packaging/package_test.go]:
```suggestion
	// only test x86 packages, excluding arm
	debs := getFiles(t, regexp.MustCompile(`(i386|i686|amd64|x86)+.*\.deb$`))
```

File: script/jenkins/package.sh [hunk 2]
```diff
@@ -1,4 +1,9 @@
 #!/usr/bin/env bash
 set -euox pipefail
 
- ./_beats/dev-tools/jenkins_release.sh
+# This controls the defaults used the Jenkins package job. They can be
+# overridden by setting them in the environement prior to running this script.
+export SNAPSHOT="${SNAPSHOT:-true}"
+export PLATFORMS="${PLATFORMS:-+linux/armv7 +linux/ppc64le +linux/s390x +linux/mips64}"
```
Comment (graphaelli) [inline script/jenkins/package.sh]: This is a good find, why do we even bother to build for platforms we don't support (yet)?

File: tests/packaging/package_test.go (hunk 1 above)
Comment (graphaelli) [reply tests/packaging/package_test.go]: If we change the platforms we build for we wouldn't need this

File: script/jenkins/package.sh (hunk 2 above)
Comment (kuisathaverat) [reply script/jenkins/package.sh]: it is what the Makefile uses, we can remove all non-x86 packages

File: Jenkinsfile [hunk 3]
```diff
@@ -419,7 +416,7 @@ pipeline {
             branch "v\\d?"
             tag "v\\d+\\.\\d+\\.\\d+*"
             expression { return params.Run_As_Master_Branch }
-            expression { return env.BEATS_UPDATED != "0" }
+            expression { return env.BEATS_UPDATED != "false" }
```
Comment (kuisathaverat) [inline Jenkinsfile]: fix build trigger, it was trigger always not only with beats updates.

File: Jenkinsfile [hunk 4]
```diff
@@ -57,7 +57,7 @@ pipeline {
         stash allowEmpty: true, name: 'source', useDefaultExcludes: false
         script {
           dir("${BASE_DIR}"){
-            env.GO_VERSION = readFile(".go-version")
+            env.GO_VERSION = readFile(".go-version").trim()
```
Comment (kuisathaverat) [inline Jenkinsfile]: remove \n

File: Jenkinsfile [hunk 5]
```diff
@@ -447,8 +447,17 @@ pipeline {
...
   }
   post {
-    always {
+    cleanup {
       notifyBuildResult()
     }
   }
 }
+
+def golang(Closure body){
```
Comment (kuisathaverat) [inline Jenkinsfile]: launch a Docker container with Go and Docker support

File: magefile.go [hunk 6]
```diff
@@ -127,7 +127,7 @@ func TestPackagesInstall() error {
 	args = append(args, "-tags=package")
 
 	if out, err := goTest(args...); err != nil {
-		if !mg.Verbose() {
+		if mg.Verbose() {
```
Comment (kuisathaverat) [inline magefile.go]: the condition was inverted, now show the messages on verbose mode

File: tests/packaging/package_test.go [hunk 7]
```diff
@@ -44,7 +45,7 @@ func TestDeb(t *testing.T) {
 }
 
 func TestRpm(t *testing.T) {
-	rpms := getFiles(t, regexp.MustCompile(`\.rpm$`))
+	rpms := getFiles(t, regexp.MustCompile(`(i386|i686|amd64|x86)+.*\.rpm$`))
```
Comment (kuisathaverat) [inline tests/packaging/package_test.go]:
```suggestion
        // only test x86 packages, excluding arm
	rpms := getFiles(t, regexp.MustCompile(`(i386|i686|amd64|x86)+.*\.rpm$`))
```

File: .ci/docker/golang-mage/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,23 @@
+ARG GO_VERSION=1.12.4
```
Comment (kuisathaverat) [inline .ci/docker/golang-mage/Dockerfile]: Tool Docker container to run the build steps

File: Jenkinsfile [hunk 9]
```diff
@@ -130,8 +125,10 @@ pipeline {
             withGithubNotify(context: 'Build - Linux') {
               deleteDir()
               unstash 'source'
-              dir("${BASE_DIR}"){
-                sh './script/jenkins/build.sh'
+              golang(){
```
Comment (kuisathaverat) [inline Jenkinsfile]: run the build step inside the Tool Docker container

File: Jenkinsfile [hunk 10]
```diff
@@ -289,9 +284,11 @@ pipeline {
             withGithubNotify(context: 'Benchmarking') {
               deleteDir()
               unstash 'source'
-              dir("${BASE_DIR}"){
-                sh './script/jenkins/bench.sh'
-                sendBenchmarks(file: 'bench.out', index: "benchmark-server")
+              golang(){
```
Comment (kuisathaverat) [inline Jenkinsfile]: run the benchmarks step inside the Tool Docker container

File: Jenkinsfile [hunk 11]
```diff
@@ -428,8 +425,11 @@ pipeline {
         withGithubNotify(context: 'Release') {
           deleteDir()
           unstash 'source'
-          dir("${BASE_DIR}"){
-            sh './script/jenkins/package.sh'
+          golang(){
```
Comment (kuisathaverat) [inline Jenkinsfile]: run the Release step inside the Tool Docker container

File: script/jenkins/bench.sh [hunk 12]
```diff
@@ -1,9 +1,4 @@
 #!/usr/bin/env bash
 set -euxo pipefail
 
-source ./_beats/dev-tools/common.bash
-
-jenkins_setup
-
-go get -u golang.org/x/tools/cmd/benchcmp
-make bench | tee bench.out
+go test -benchmem -run=XXX -benchtime=100ms -bench='.*' ./... | tee bench.out
```
Comment (kuisathaverat) [inline script/jenkins/bench.sh]: use go command directly

File: script/jenkins/build.sh [hunk 13]
```diff
@@ -1,15 +1,5 @@
...
-
-cleanup() {
-  rm -rf $TEMP_PYTHON_ENV
-  make stop-environment fix-permissions
-}
-trap cleanup EXIT
-
-make clean
-make
+mage -debug clean
```
Comment (kuisathaverat) [inline script/jenkins/build.sh]: use mage commands directly

File: script/jenkins/package.sh [hunk 14]
```diff
@@ -1,4 +1,9 @@
...
+
+mage -debug package
```
Comment (kuisathaverat) [inline script/jenkins/package.sh]: use mage command directly

File: script/jenkins/test-install-packages.sh [hunk 15]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+set -xeuo pipefail
+
+mage -v -debug testPackagesInstall
```
Comment (kuisathaverat) [inline script/jenkins/test-install-packages.sh]: use mage command directly

**Comment by kuisathaverat** [comment]
I've added the changes to the initial commit and comments on every change

File: script/jenkins/package.sh (hunk 2 above)
Comment (graphaelli) [reply script/jenkins/package.sh]: Agreed, we should just unset this and let `packages.yml` define this instead.

**Comment by graphaelli** [comment]
🤦‍♂ I was asking for an update to the description since it didn't capture what was changing here, not commentary on each change, don't feel like you have to do that in the future.

File: script/jenkins/build.sh (hunk 13 above)
Comment (graphaelli) [reply script/jenkins/build.sh]: this isn't identical, `make clean` cleans up a bunch of other things including `build/` and `fields.yml`

**Comment by kuisathaverat** [comment]
np, It was for me also to mark the changes

File: script/jenkins/build.sh (hunk 13 above)
Comment (kuisathaverat) [reply script/jenkins/build.sh]: the workspace is clean, I can remove the clean too

**Comment by kuisathaverat** [comment]
I gonna split this in several PRs there are too many changes in the same PR

**Comment by simitt** [comment]
@kuisathaverat can you please close this PR then and link it in the smaller ones that you are going to open, or reopen once a review is desired again.

**Comment by kuisathaverat** [comment]
@simitt it is already split in several PRs, most of them merged. This PR is rebased on master branch now so there is only one unrelated change in it (elastic/apm-server#2391), if we review and merge it we can rebase this PR again for a final review.

* add labels to script steps - elastic/apm-server#2378
* add Tool Docker container with the Golang version used and tools needed - elastic/apm-server#2375
* use plain Go or mage commands inside the Tool Docker container - elastic/apm-server#2375
* fix Release build trigger, it was trigger always not only with beats updates. - elastic/apm-server#2379
* move benchmark stage to run in the tool Docker container - elastic/apm-server#2391

**Comment by graphaelli** [comment]
Also out of this came:
* build default packages #2368 
* vendor check/test tools #2395

**Comment by simitt** [comment]
@graphaelli I leave this review up to you, as I think you've been more involved in the mage changes than I was.

File: tests/packaging/package_test.go [hunk 16]
```diff
@@ -34,7 +34,8 @@ var (
 )
 
 func TestDeb(t *testing.T) {
-	debs := getFiles(t, regexp.MustCompile(`\.deb$`))
+	// only test x86 packages, excluding arm
+	debs := getFiles(t, regexp.MustCompile(`(i386|i686|amd64|x86)*.\.deb$`))
```
Comment (graphaelli) [inline tests/packaging/package_test.go]: Let's revert this as the jenkins build doesn't need it any more

File: tests/packaging/package_test.go (hunk 7 above)
Comment (graphaelli) [reply tests/packaging/package_test.go]: Let's revert this as the jenkins build doesn't need it any more

**Pull request merged**